### PR TITLE
Pm 27900 add additional hardening in extension frame validation

### PR DESCRIPTION
**Pull request by dan-livefront**
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-27900

## 📔 Objective

Help better secure the extension’s inline menu frame communication logic with additional hardening and validation. The goal is to improve security, reliability, and resilience in how the extension handles frame interactions.

Key Improvements

- Introduces stricter validation of incoming frame messages to verify structure and data integrity before processing.
- Adds safeguards to detect and reject malformed, unexpected, or potentially malicious frame data.
- Refines the logic governing allowed message origins and data types, reducing the risk of unauthorized communications.
- Ensures only valid and expected frame messages are processed, strengthening the extension’s communication layer against misuse and edge case exploits.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts [hunk 1]
```diff
@@ -97,6 +151,18 @@ export class AutofillInlineMenuContainer {
...
+      this.postMessageToInlineMenuPageUnsafe(messageWithToken);
+    }
+  }
+
+  /**
+   * posts a message to the inline menu page iframe without token validation, used internally when we need to send raw postMessage.
+   *
+   * @param message - The message to post.
+   */
+  private postMessageToInlineMenuPageUnsafe(message: Record<string, unknown>) {
```
Comment (dan-livefront) [inline apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts:169]: potentially make token required and change this to only be something like     `this.inlineMenuPageIframe.contentWindow.postMessage(messageWithToken, this.extensionOrigin);`

File: apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts [hunk 2]
```diff
@@ -97,6 +151,22 @@ export class AutofillInlineMenuContainer {
...
+   *
+   * UNSAFE: Bypasses token authentication and sends raw messages. Only use internally
+   * when sending trusted messages (e.g., initialization) or when token validation
+   * would create circular dependencies. External callers should use postMessageToInlineMenuPage().
+   *
+   * @param message - The message to post.
+   */
+  private postMessageToInlineMenuPageUnsafe(message: Record<string, unknown>) {
     if (this.inlineMenuPageIframe?.contentWindow) {
       this.inlineMenuPageIframe.contentWindow.postMessage(message, "*");
```
[the remark this answers is not in the document]
Comment (dan-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts:171]: The iframe is sandboxed with only allow-scripts, so it has a null origin. When it posts to the parent, using a specific origin can be blocked. The container already uses object identity verification and token auth, so origin checking isn't needed for iframe messages

File: apps/browser/src/autofill/overlay/inline-menu/pages/shared/autofill-inline-menu-page-element.ts [hunk 3]
```diff
@@ -58,7 +63,11 @@ export class AutofillInlineMenuPageElement extends HTMLElement {
    * @param message - The message to post
    */
   protected postMessageToParent(message: AutofillInlineMenuPageElementWindowMessage) {
-    globalThis.parent.postMessage({ portKey: this.portKey, ...message }, "*");
+    const messageWithAuth: Record<string, unknown> = { portKey: this.portKey, ...message };
+    if (this.token) {
+      messageWithAuth.token = this.token;
+    }
+    globalThis.parent.postMessage(messageWithAuth, "*");
```
[the remark this answers is not in the document]
Comment (dan-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/shared/autofill-inline-menu-page-element.ts:70]: The iframe is sandboxed with only allow-scripts, so it has a null origin. When it posts to the parent, using a specific origin can be blocked. The container already uses object identity verification and token auth, so origin checking isn't needed for iframe messages

File: apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts [hunk 4]
```diff
@@ -49,11 +72,8 @@ export class AutofillInlineMenuContainer {
   };
 
   constructor() {
-    this.extensionOriginsSet = new Set([
-      chrome.runtime.getURL("").slice(0, -1).toLowerCase(), // Remove the trailing slash and normalize the extension url to lowercase
-      "null",
-    ]);
-
+    this.token = generateRandomChars(32);
```
[the remark this answers is not in the document]
Comment (jholland-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts:75]: I'm going to resolve this because under the hood we are using `crypto.getRandomValues()`

File: apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts [hunk 5]
```diff
@@ -184,10 +270,28 @@ export class AutofillInlineMenuContainer {
     if (!this.inlineMenuPageIframe) {
       return false;
     }
+    // only trust the specific iframe we created
+    return this.inlineMenuPageIframe.contentWindow === event.source;
+  }
+
+  private verifyToken(message: { token?: string }): boolean {
```
[the remark this answers is not in the document]
Comment (jholland-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts]: I think the function name may be causing this Claude comment.  We aren't verifying the token cryptographically.  We are checking that the tokens match.  I would suggest renaming this.

[the remark this answers is not in the document]
Comment (dan-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/menu-container/autofill-inline-menu-container.ts]: Great call out Jeff, updated the name and improved documentation for clarity. Further clarifying here for The Claude 😂 These tokens are already short lived since the inline menu regenerates one whenever it’s reinitialized such as when any command is used and the inline menu is closed. Adding a nonce for critical commands provides no additional security because the token already serves that purpose.
